pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go

```diff
@@ -0,0 +1,123 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package mixedversion
+
+import "math/rand"
+
+const (
+	// PreserveDowngradeOptionRandomizer is a mutator that changes the
+	// timing in which the `preserve_downgrade_option` cluster setting
+	// is reset during the upgrade test. Typically (when this mutator is
+	// not enabled), this happens at the end of the `LastUpgrade`
+	// stage, when all nodes have been restarted and are running the
+	// target binary version. However, we have observed bugs in the past
+	// that only manifested when this setting was reset at different
+	// times in the test (e.g., #111610); this mutator exists to catch
+	// regressions of this type.
+	PreserveDowngradeOptionRandomizer = "preserve_downgrade_option_randomizer"
+)
+
+type preserveDowngradeOptionRandomizerMutator struct{}
+
+func (m preserveDowngradeOptionRandomizerMutator) Name() string {
+	return PreserveDowngradeOptionRandomizer
+}
+
+// Most runs will have this mutator disabled, as the base upgrade
+// plan's approach of resetting the cluster setting when all nodes are
+// upgraded is the most sensible / common.
+func (m preserveDowngradeOptionRandomizerMutator) Probability() float64 {
+	return 0.3
+}
+
+// Generate returns mutations to remove the existing step to reset the
+// `preserve_downgrade_option` cluster setting, and reinserts it back
+// in some other point in the test, before all nodes are upgraded. Not
+// every upgrade in the test plan is affected, but the upgrade to the
+// current version is always mutated.
+func (m preserveDowngradeOptionRandomizerMutator) Generate(
+	rng *rand.Rand, plan *TestPlan,
+) []mutation {
+	var mutations []mutation
+	for _, upgradeSelector := range randomUpgrades(rng, plan) {
+		removeExistingStep := upgradeSelector.
+			Filter(func(s *singleStep) bool {
+				_, ok := s.impl.(allowUpgradeStep)
+				return ok
+			}).
+			Remove()
+
+		addRandomly := upgradeSelector.
+			Filter(func(s *singleStep) bool {
+				// It is valid to reset the cluster setting when we are
+				// performing a rollback (as we know the next upgrade will be
+				// the final one); or during the final upgrade itself.
+				return (s.context.Stage == LastUpgradeStage || s.context.Stage == RollbackUpgradeStage) &&
+					// We also don't want all nodes to be running the latest
+					// binary, as that would be equivalent to the test plan
+					// without this mutator.
+					len(s.context.NodesInNextVersion()) < len(s.context.CockroachNodes)
+			}).
+			RandomStep(rng).
+			// Note that we don't attempt a concurrent insert because the
+			// selected step could be one that restarts a cockroach node,
+			// and `allowUpgradeStep` could fail in that situation.
+			InsertBefore(allowUpgradeStep{})
+
+		// Finally, we update the context associated with every step where
+		// all nodes are running the next verison to indicate they are in
+		// fact in `Finalizing` state. Previously, this would only be set
```

Comment:
I'm not sure I understand the question, so I'll go for a longer answer in hopes it will cover what you are asking.

Before any test runs, we generate a "test plan" for the upgrade test. This includes the specific ordering of every event in the test: when nodes will restart, when user provided functions will be called, etc. The mixedversion context contains information relative to the mixed-version state the cluster is in, and is passed to user functions so that they can adapt based on current cluster state (e.g., pick a node running the older/newer version, do something different if migrations are running in the background, etc). All randomization happens during planning.

The mutators concept was added later, and is technically still part of the planning stage. Once we generate a base test plan, we non-deterministically apply a subset of the available mutators. When the mutator is chosen, we call `Generate` on the corresponding implementation. The mutations are then applied to the test plan, and the resulting test plan is passed to the runner, which blindly runs the steps in the test plan.

You probably already knew all of that, but I wanted to make sure we were on the same page. Specifically, I'm confused by the phrase

> `allowUpgradeStep` is guaranteed to run after the mutator

as `allowUpgradeStep` does _run_ but a mutator doesn't "run" in the same way -- it just changes the test plan.

`allowUpgradeStep` is a step in the test plan that resets `cluster.preserve_downgrade_option`. The time it runs is dependent on where the planner decides to place it in the test plan. By default (when the mutator introduced in this PR is not applied), we always reset once every node is restarted and is running the binary we want to upgrade to. When this mutator is enabled, we might reset it at different times (e.g., during a rollback, or halfway during our upgrade process).

When the mutator is _not_ enabled, we know that upgrade migrations will only run once `allowUpgradeStep` finishes. However, when the mutator _is_ enabled, migrations will run "as soon as all nodes are are on the next version". This is what the context update is capturing. If we didn't perform this update in the context, we would have steps that mistakenly thought they were _not_ finalizing when in fact they are. For a concrete example of this, see the thread above with Darryl and step 71 in the test added in this PR:

https://github.com/cockroachdb/cockroach/blob/2ae25505ce5edb4fbae9035abef9fe52f8ea5dfe/pkg/cmd/roachtest/roachtestutil/mixedversion/testdata/planner/preserve_downgrade_option_randomizer#L103-L108

Step 71 only knows it's "finalizing" because of this code that updates the context now that the `allowUpgradeStep` was moved.

*****

Hope some of this makes sense. If I misunderstood the question, happy to try to answer again.

Reply:
> When the mutator is not enabled, we know that upgrade migrations will only run once allowUpgradeStep finishes. However, when the mutator is enabled, migrations will run "as soon as all nodes are are on the next version"

Right. I think what I was missing is the fact that `allowUpgradeStep` doesn't set (or propagate) `Finalizing`. This happens in `finalizeUpgradeSteps`, which propagates `Finalizing=true` into the next steps. Since the mutator inserts a sequence of upgrade steps after `Plan`, `Finalizing=true` isn't propagated for this (new) sequence (since `finalizeUpgradeSteps` isn't invoked in this case). 

Note, if we had the knowledge all (transitive) dependents of `allowUpgradeStep`, then this context propagation could be moved directly into the step as opposed to being exposed to a mutator. More generally, we might want to think about what it would take to make _dependent_ sequences of steps, where the (dependence) ordering "closes" the whole sequence, i.e., invariants over the whole sequences are preserved.

In any case, thanks for a very detailed explanation!

Reply:
>  I think what I was missing is the fact that allowUpgradeStep doesn't set (or propagate) Finalizing.

Yeah, this would be tricky, as it wouldn't be safe to set `Finalizing` during `allowUpgradeStep` -- we might not be finalizing right after that step. `Finalizing` is a property of the mixed-version state, and `allowUpgradeStep` is only a part of that formula.

> then this context propagation could be moved directly into the step as opposed to being exposed to a mutator

IMO, a lot of the advantages of the current mutator API come from the fact that it _does_ have access to the context. For instance, in the mutator added in this PR, it's very easy to say "insert the allowUpgradeStep in any step during the rollback or last upgrade stages, and also choose a point in time when not all nodes are running the new binary". The context is what allows us to "query" the mixed-version state in this way.

But I also see your point and I think there would be advantages to steps managing the context. As I said above as well, I don't see us having to update the context in the vast majority of mutators so I'm not too concerned about this particular issue. We'll see if that changes.